lib/collection/set.js

```diff
@@ -41,31 +39,35 @@ export class SetColl extends Collection {
   }
 
   remove(item) {
-    this._removeWithoutObserver(item);
-    this._notifyRemoved([item]);
+    if (this._removeWithoutObserver(item)) {
+      this._notifyRemoved([item]);
+    }
   }
 
   _removeWithoutObserver(item) {
-    this._set.delete(item);
+    return this._set.delete(item);
   }
 
   addAll(items) {
-    for (let item of items) {
-      this._addWithoutObserver(item);
+    items = items.filter(item => this._addWithoutObserver(item));
```

Comment:
Are you aware that `items` can be a `Collection` object? And that `coll.filter()` will create a dynamic filter object, with observers and all? We should not use that for primitive operations.

Please use the fasted, most light-weight operations (which are still readable) for basic operations.

Function calls are OK (e.g. `this.hasItems` instead of `this._set.size != 0`), and are encouraged for readability and reduction of code duplicaltion, but you need to know what they do.

**IMPORTANT**: Same for all the other functions (which you changed with the purpose of reducing the notifications to the actually removed values).

Reply:
I remember now: I was getting collections passed to notifications, which really messes things up when you're trying to do remote notifications.

Reply:
A `for` loop creates only one array.

(same for remove below)